### ISSUE TITLE
1.18.x extension build failed

### ISSUE THREAD

**Issue by vladimirtemnikov** [Shopify/bootsnap#470]
Hello, latest version's ext can't compile.
Specs: MacOS 14.3, Apple silicon, Ruby 3.2.3
1.17.1 compiles well.

<details>
  <summary>mkmf.log</summary>

  ```
have_func: checking for fdatasync() in fcntl.h... -------------------- yes

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
conftest.c:16:57: error: use of undeclared identifier 'fdatasync'
   16 | int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
      |                                                         ^
1 error generated.
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <fcntl.h>
 4: 
 5: /*top*/
 6: extern int t(void);
 7: int main(int argc, char **argv)
 8: {
 9:   if (argc > 1000000) {
10:     int (* volatile tp)(void)=(int (*)(void))&t;
11:     printf("%d", (*tp)());
12:   }
13: 
14:   return !!argv[argc];
15: }
16: int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
/* end */

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <fcntl.h>
 4: 
 5: /*top*/
 6: extern int t(void);
 7: int main(int argc, char **argv)
 8: {
 9:   if (argc > 1000000) {
10:     int (* volatile tp)(void)=(int (*)(void))&t;
11:     printf("%d", (*tp)());
12:   }
13: 
14:   return !!argv[argc];
15: }
16: extern void fdatasync();
17: int t(void) { fdatasync(); return 0; }
/* end */

--------------------

block in append_cppflags: checking for whether _GNU_SOURCE is accepted as CPPFLAGS... -------------------- no

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe  _GNU_SOURCE -Werror -c conftest.c"
clang: error: no such file or directory: '_GNU_SOURCE'
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

--------------------

block in append_cflags: checking for whether -O3 is accepted as CFLAGS... -------------------- yes

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe  -O3 -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

--------------------

block in append_cflags: checking for whether -std=c99 is accepted as CFLAGS... -------------------- yes

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe -O3  -std=c99 -Werror -c conftest.c"
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

--------------------

  ```
</details>

```
compiling bootsnap.c
bootsnap.c:343:7: error: call to undeclared function 'fdatasync'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  343 |   if (fdatasync(cache_fd) < 0) {
      |       ^
1 error generated.
make: *** [bootsnap.o] Error 1
```

**Comment by vladimirtemnikov** [comment]
@casperisfine thank you for quick response.
Trying to figure out what's wrong.

Strange that conftest fails literally below check for fdatasync:

```
DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
conftest.c:16:57: error: use of undeclared identifier 'fdatasync'
 16 | int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
    |                                                         ^
1 error generated.
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: #include <fcntl.h>
4: 
5: /*top*/
6: extern int t(void);
7: int main(int argc, char **argv)
8: {
9:   if (argc > 1000000) {
10:     int (* volatile tp)(void)=(int (*)(void))&t;
11:     printf("%d", (*tp)());
12:   }
13: 
14:   return !!argv[argc];
15: }
16: int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
/* end */
```

**Comment by vladimirtemnikov** [comment]
Adding `extern int fdatasync(int);` before calling fixes the problem, but I think it is more crutch than real solve (I'm not familiar with building exts).

**Comment by vladimirtemnikov** [comment]
@casperisfine used main branch, still does not work:
```
have_func: checking for fdatasync() in unistd.h... -------------------- yes

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
checked program was:
/* begin */
1: #include "ruby.h"
2: 
3: int main(int argc, char **argv)
4: {
5:   return !!argv[argc];
6: }
/* end */

DYLD_FALLBACK_LIBRARY_PATH=.:/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib "clang -o conftest -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/arm64-darwin23 -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0/ruby/backward -I/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/include/ruby-3.2.0 -I. -D_XOPEN_SOURCE -D_DARWIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT   -fdeclspec -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wextra-tokens -Wdeprecated-declarations -Wdivision-by-zero -Wdiv-by-zero -Wimplicit-function-declaration -Wimplicit-int -Wmisleading-indentation -Wpointer-arith -Wshorten-64-to-32 -Wwrite-strings -Wold-style-definition -Wmissing-noreturn -Wno-cast-function-type -Wno-constant-logical-operand -Wno-long-long -Wno-missing-field-initializers -Wno-overlength-strings -Wno-parentheses-equality -Wno-self-assign -Wno-tautological-compare -Wno-unused-parameter -Wno-unused-value -Wunused-variable -Wundef  -fno-common -pipe conftest.c  -L. -L/Users/vladimirtemnikov/.local/share/mise/installs/ruby/3.2.3/lib -L. -fstack-protector-strong -L/opt/homebrew/opt/gmp/lib     -lruby.3.2  -lpthread  "
conftest.c:16:57: error: use of undeclared identifier 'fdatasync'
   16 | int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
      |                                                         ^
1 error generated.
checked program was:
/* begin */
 1: #include "ruby.h"
 2: 
 3: #include <unistd.h>
 4: 
 5: /*top*/
 6: extern int t(void);
 7: int main(int argc, char **argv)
 8: {
 9:   if (argc > 1000000) {
10:     int (* volatile tp)(void)=(int (*)(void))&t;
11:     printf("%d", (*tp)());
12:   }
13: 
14:   return !!argv[argc];
15: }
16: int t(void) { void ((*volatile p)()); p = (void ((*)()))fdatasync; return !p; }
/* end */
```

**Comment by casperisfine** [comment]
That is wild.

**Comment by casperisfine** [comment]
Alright, if PHP and PG do it, I don't see why not use that workaround. Fixed in d4b46660cb70765879286bd8d9af51ce05cbb021
